ggv_generator.py

```diff
@@ -13,20 +14,43 @@ def ggv_generator(vehicle_params, sweep_ranges:dict, mesh:int):
     steered_angle_sweep = np.linspace(sweep_ranges["steered_angle"][0], sweep_ranges["steered_angle"][1], mesh)
     torque_request = np.linspace(sweep_ranges["torque_request"][0], sweep_ranges["torque_request"][1], mesh)
     is_left_diff_bias = sweep_ranges["is_left_diff_bias"]
+    vehicles = [vehicle_params]
 
-    ### ~~~ MULTIPROCESSING BELOW ~~~ ###
+    ### ~~~ SOLVER BELOW ~~~ ###
+    states_product = itertools.product(vehicles, body_slip_sweep, steered_angle_sweep, s_dot_sweep, torque_request, is_left_diff_bias)
     t1 = perf_counter()
-    p = multiprocessing.Pool(multiprocessing.cpu_count())
-    state_solver2 = lambda x: engine.Solver(vehicle_params).solve(engine.State(*x))
-    states_product = itertools.product(body_slip_sweep, steered_angle_sweep, s_dot_sweep, torque_request, is_left_diff_bias)
-    return_list = tqdm(p.imap(state_solver2, states_product))
-    p.close()
-
-    ### ~~~ EXPORT MMM RESULTS ~~~ ###
-    log_df = pd.DataFrame.from_records(filter(None, return_list))
-    log_df.to_csv("analysis/MMM.csv")
-    print(f"Sweep completed in {int(perf_counter() - t1)} seconds, exported to CSV")
-
-# def state_solver(inputs):
-#     vehicle, state_inputs = inputs[:1], inputs[1:]
-#     return engine.Solver(vehicle_params.EasyDriver()).solve(engine.State(*state_inputs))
\ No newline at end of file
+    with multiprocessing.Pool(multiprocessing.cpu_count()) as p:
+        results = tqdm(p.imap_unordered(solve_wrapper, states_product))
+        df = pd.DataFrame(filter(None, results))
+    print(f"Sweep completed in {int(perf_counter() - t1)} seconds")
+
+    ### ~~~ FILTERING ~~~ ###
+    tires = ["front_left", "front_right", "rear_left", "rear_right"]
+    filt_df = df[(df["roll"]*180/np.pi < 5) & (df["yaw_acceleration"] < 200) & (df["motor_angular_velocity"] < vehicle_params.max_motor_speed)]
+    for tire in tires:
+        filt_df = filt_df[(filt_df[f"{tire}_tire_is_saturated"] == False)
+                 & (filt_df[f"{tire}_tire_tire_centric_forces_2"] > 1) & (filt_df[f"{tire}_tire_tire_centric_forces_2"] < 4000)]
+    # Filter if power limit hit
+    # TOD mapping
+    filt_df = filt_df[filt_df["motor_angular_velocity"] * filt_df["motor_torque"] < 77000]
+
+    # 80 kW = P = W / t = F * d/t = F * v = T * w = (230 * 3.85) *
+    # P
+    # P = I * V = (T / Kt) * V
+
+    ### ~~~ CONVEX HULL ~~~ ###
+    hull_df = pd.DataFrame()
+    for vel in filt_df["s_dot"].unique():
```

Comment:
I have a concern on the G-G diagrams i've been seeing this producing. The current form seems to imply that there's a concave section of the G-G at the negative long accel cases, but we just convex hull over it. There's a huge region of unattainable states, per our MMM, that are being considered for the lapsim as a result. One of these is true:

1. Our MMM generates an incorrect set of states and a bug exists somewhere else
2. GG slices of the GGV are not inherently convex, and using a convexhull is misleading when computing laptimes

Reply:
Agreed with this, seems like a long term issue. Will add it to list